### PR TITLE
Diagonal/diagonalize functions

### PR DESCRIPTION
**Pull request by pettni**
Additions discussed in #288

### CODE REVIEW

File: sparse/_coo/core.py [hunk 1]
```diff
@@ -2258,6 +2258,112 @@ def asformat(self, format, compressed_axes=None):
...
+        :obj:`numpy.diagonal`: NumPy equivalent function
+        """
+        if self.shape[axis1] != self.shape[axis2]:
+            raise ValueError('shape[axis1] != shape[axis2]')
+
+        diag_axes = [axis for axis in range(len(self.shape)) if axis != axis1 and axis != axis2] + [axis1]
+        diag_shape = [self.shape[axis] for axis in diag_axes]
+        diag_shape[-1] -= abs(offset)
+
+        diag_idx = [i for i in range(len(self.data)) if self.coords[axis1][i]+offset == self.coords[axis2][i]]
```
Comment (hameerabbasi) [inline sparse/_coo/core.py]: Is it possible to move this loop into Numba?

File: sparse/_coo/core.py [hunk 2]
```diff
@@ -2258,6 +2258,112 @@ def asformat(self, format, compressed_axes=None):
 
         raise NotImplementedError("The given format is not supported.")
 
+    def diagonal(self, offset=0, axis1=0, axis2=1):
```
Comment (hameerabbasi) [inline sparse/_coo/core.py]: Let's convert this into functions rather than methods, as it is in NumPy.

File: sparse/_coo/core.py [hunk 3]
```diff
@@ -2258,6 +2258,112 @@ def asformat(self, format, compressed_axes=None):
...
+        diag_shape[-1] -= abs(offset)
+
+        diag_idx = [i for i in range(len(self.data)) if self.coords[axis1][i]+offset == self.coords[axis2][i]]
+
+        diag_coords = [self.coords[axis][diag_idx] for axis in diag_axes]
+        diag_data  = self.data[diag_idx]
+
+        return COO(diag_coords, diag_data, diag_shape)
+
+    def diagonalize(self, axis=0):
```
Comment (hameerabbasi) [inline sparse/_coo/core.py]: Method -> Function here as well. Also need to mirror the NumPy API...

Comment (pettni) [reply sparse/_coo/core.py]: As discussed in #288, there is no numpy equivalent of ```diagonalize```. Thereis ```np.diag``` but that only does the same thing for 1D and actually does the opposite for 2D. It does not accept 3D or higher. I thought calling it ```diagonalize``` was the best solution to not confuse

Comment (hameerabbasi) [reply sparse/_coo/core.py]: That seems reasonable... But future code might not work if we have to change the signature to be in line with NumPy, so I think a warning is in order in the docs.

File: sparse/_coo/common.py [hunk 4]
```diff
@@ -1215,6 +1215,127 @@ def ones_like(a, dtype=None):
...
+    diag_data = a.data[diag_idx]
+
+    return COO(diag_coords, diag_data, diag_shape)
+
+
+def diagonalize(a, axis=0):
+    """
+    Diagonalize a COO array. The new dimension is appended at the end.
+
+    **Warning**: :obj:`diagonalize` is not :obj:`numpy` compatible as there is no direct :obj:`numpy` equivalent. The API may change in the future.
```
Comment (hameerabbasi) [inline sparse/_coo/common.py]: This should be in a [`..warning::` directive](http://docutils.sourceforge.net/docs/ref/rst/directives.html#warning).

**Comment by hameerabbasi** [comment]
You can run `black .` to fix the linting errors.

**Comment by hameerabbasi** [comment]
Excellent, thanks, @pettni!

**Comment by pettni** [comment]
Happy to contribute, thanks for the quick reviews!

**Pull request merged**